Extend teal transform vignette (#1301)
insightsengineering/teal: vignettes/data-transform-as-shiny-module.Rmd

---
title: "Data Transformations as shiny Module"
author: "NEST CoreDev"
output: 
  rmarkdown::html_vignette:
    toc: true
vignette: >
  %\VignetteIndexEntry{Data Transformations as shiny Module}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

## Introduction

`teal` ver `0.15.x` (TO BE UDPATED) introduced new argument in `teal::module` called `transformers`. 

> This argument allows to pass a `list` of `teal_data_module` objects that are created either with `teal_transform_module()` or `teal_data_module()` functions.

Reviewer: I'm unsure if it's necessary to mention using the `teal_data_module()` function to create the object and pass it to the `transformers` argument. This may make sense from a developer's perspective since `teal_transform_module()` also returns a `teal_data_module` class, but I'm not sure if it's useful information for the user.

In all the examples below, we only use `teal_transform_module()` to create the object passed to `transformers`.

If we decide to keep the mention of `teal_data_module()`, we would need to include examples of how to pass the object to `transformers` using that function. I expect that using `teal_data_module()` would involve additional steps, which might not be ideal since `teal_transform_module()` already handles this.

At the moment, I prefer to focus on `teal_transform_module()`, but I'd like to hear your thoughts on this.

```suggestion
This argument allows to pass a `list` of `teal_data_module` objects that are created using `teal_transform_module()` function.
```

Reviewer 2: Agree. I also had those thoughts when I wrote this sentence. Will accept the change

Reviewer 2: Actually @vedhav can you accept during your last review? This all is ready to be accepted and merged

Author: Good catch Dony, It's actually only important to talk about the object `teal_data_module` in the context of the transform module to make sure that we communicate that both the DDL and transform use the same object.
But, right now we already have different features between the object created with `teal_transform_module()` and `teal_data_module()` that it does not make sense to use the later for transforming.

The main benefit of `teal_transform_module()` is the ability to transform data before passing the 
data into the module. This feature allows to extend the regular behavior of existing modules by specifying custom data operations on data inside this module.

`teal_transform_module()` is a Shiny module that takes `ui` and `server` arguments. When provided, `teal` will execute data transformations for the specified module when it is loaded and each time the data is changed. `server` extend the logic behind data manipulations, where `ui` extends filter panel with new UI elements that orchestrate the transformer inputs.

This vignette presents the way on how to manage custom data transformations in `teal` apps.

## Creating your first custom data transformation module

We initialize a simple `teal` app where we pass `iris` and `mtcars` as the input datasets.

```{r, message = FALSE, warning = FALSE}

> library(teal)

Reviewer: Since the `teal` library is already loaded in an earlier section, there's no need to repeat `library(teal)` in this chunk or in any subsequent chunks. WDYT?

Author: Thanks, removed instances of redundant library calls.

When a custom transformation is used, the UI for the transformation is placed below the filter panel.
However, there is a way to customize the placement of the UI inside the module content.

In order to place the transformation UI inside the module there are few things one has to do:
1. Create a custom module wrapper function.

> 2. In the module wrapper function, Call the module and store it in a variable.

Author: ```suggestion
2. Call the desired module in the module wrapper function and store it in a variable so it can be modified.
```

> 2. Call the desired module in the module wrapper function and store it in a variable so it can be modified.

Reviewer 2: ```suggestion
2. Call the desired module in the module wrapper function and store it in a variable so it's UI can be modified.
```

> 3. Modify the UI of the module with the transform UI at the desired location by calling the `ui_transform_data`. Note that in order for the transform to work you need to change the name space of the id by passing `NS(gsub("-module$", "", id), "data_transform")`.

Author: @averissimo The namespace change is the part that is not great. Is there another way to do this?
@m7pr suggested to have some kind of a wrapper to achieve this. Either for the name space or the whole module customization itself.

Reviewer 3: I think there needs to be a trick somewhere as we're injecting the `ui_transform` under different namespace locations, so we have 2 options:

1. As is with no namespace trick when injecting the `ui_transform_data` in most cases and trick on manual
2. The other way round

From a design point of view, it seems that the trick should be applied on the exception/rare case. But this makes it module-developer visible.

Feel more than free to propose a change :-D ... I'm not keen on it as it's a lesser of evils kind of implementation.

I've tried getting to the `SessionProxy` parent, but the public API seems to not allow for it, `override` of `SessionProxy` is lost as scope is created (only accessing the original session)

Reviewer 2: Would this help a bit, if we expose/create such a function in this example?
```
transform_id <- function(id) {
   NS(gsub("-module$", "", id), "data_transform")
}
```

Author: `transform_id` does simplify getting the namespace with the added advantage that we can change the namespaces in teal and still not break apps using this feature because they use our wrapper.
Are we okay to export a function that caters to a small audience? Advanced users should be okay with this kind of complexity, moreover, we will have teal transform wrappers and this could be a topic to discuss during that time.
And this feature could be like a hidden feature that we don't advertise.

> 3. Modify the UI of the module with the transform UI at the desired location by calling the `ui_transform_data`. Note that in order for the transform to work you need to change the name space of the id by passing `NS(gsub("-module$", "", id), "data_transform")`.

Reviewer 2: ```suggestion
3. Modify the UI of the module with the transform UI at the desired location by calling the `ui_transform_data`. Note that in order for the transform to work you need to change the namespace of the `id` by passing `NS(gsub("-module$", "", id), "data_transform")`.
```

4. Set the `custom_ui` attribute of the `module$transformers` to `TRUE`.

Now the custom module should embed the transformation UI inside the module content.

Here is an example of a custom module wrapper function that modifies the `example_module` module.
```{r}
example_module_encoding <- function(label = "example module (on encoding)", datanames = "all", transformers = list()) {
  mod <- example_module(label, datanames, transformers)
  mod$ui <- function(id) {
    ns <- NS(id)
    teal.widgets::standard_layout(
      output = verbatimTextOutput(ns("text")),
      encoding = tags$div(
        ui_transform_data(NS(gsub("-module$", "", id), "data_transform"), transformers),
        selectInput(ns("dataname"), "Choose a dataset", choices = NULL),
        teal.widgets::verbatim_popup_ui(ns("rcode"), "Show R code")
      )

>     )

Reviewer 2: Maybe we can invent an example that is not based on `example_module`?
Maybe you can create a simple `module` that uses `ui` and `server` and you can show an application on whether it uses the default placement, and a custom placement?

This way it might be a bit more clear, instead of showing an example where we use pre-defined module and then we adjust it's UI. Could be simpler to have simple module with ui and server and then version adjusted with ui_transform_data added

Author: I wanted to do it as you mentioned by creating a custom module and then injecting this.
However, The typical use case for this would be to inject the custom transform in a module from a package like tmc. And, the vignette would become longer with custom module code which might distract the users from the actual important code needed to position the transform.
So, I decided to just place the transform to the custom location from the example mentioned above.

Reviewer 2: hmmm, I'll think about it

Reviewer 2: I think it is fine to use `example_module` if we got with the wrapper that I proposed

  }

>   attr(mod$transformers, "custom_ui") <- TRUE

Reviewer 2: Trying to figure out if this can be added automatically
